akka-stream/src/main/scala/akka/stream/stage/GraphStage.scala

```diff
@@ -984,31 +997,176 @@ abstract class GraphStageLogic private[stream] (val inCount: Int, val outCount:
+          }
+        } else done(promise)
+        //fail promise as stream is completed
+        case Completed ⇒ done(failPromiseOnComplete(promise))
+      }
+    }
+
+    private def onPending(status: Pending, promise: Promise[Done], event: T): TailRec[Promise[Done]] = {
+      if (!currentState.compareAndSet(status, Pending(Event(event, OptionVal(promise)) :: status.pendingEvents))) {
+        concurrentSet.remove(promise)
+        tailcall(invokeWithPromise(event))
+      } else done(promise)
+    }
+
+    private def failPromiseOnComplete(promise: Promise[Done]): Promise[Done] = {
+      concurrentSet.remove(promise)
+      promise.tryFailure(new StreamDetachedException())
+      promise
+    }
+
+    //external call
+    override def invoke(event: T): Unit = {
+      @tailrec
+      def internalInvoke(event: T): Unit = currentState.get() match {
+        //not started yet
+        case list @ Pending(l) ⇒ if (!currentState.compareAndSet(list, Pending(Event(event, OptionVal.None) :: l))) internalInvoke(event)
+        //started - can just send message to stream
+        case Initialized       ⇒ onAsyncInput(event)
+        //initializing is in progress in another thread (initializing thread is managed by akka)
+        case Initializing ⇒ if (!currentState.compareAndSet(Initializing, Pending(Event(event, OptionVal.None) :: List.empty[Event]))) {
+          (currentState.get(): @unchecked) match {
+            case list @ Pending(l) ⇒ if (!currentState.compareAndSet(list, Pending(Event(event, OptionVal.None) :: l))) internalInvoke(event)
+            case Initialized       ⇒ onAsyncInput(event)
+          }
+        }
+        case Completed ⇒ //do nothing here as stream is completed
+      }
+      internalInvoke(event)
     }
   }
 
   /**
    * Java API: Obtain a callback object that can be used asynchronously to re-enter the
    * current [[GraphStage]] with an asynchronous notification. The [[invoke()]] method of the returned
-   * [[AsyncCallback]] is safe to be called from other threads and it will in the background thread-safely
-   * delegate to the passed callback function. I.e. [[invoke()]] will be called by the external world and
+   * [[AsyncCallback]] is safe to be called from other threads. It will in the background thread-safely
+   * delegate to the passed callback function. I.e. [[invoke()]] will be called by other thread and
    * the passed handler will be invoked eventually in a thread-safe way by the execution environment.
    *
+   * [[AsyncCallback.invokeWithFeedback()]] has an internal promise that will be failed if event cannot be processed due to stream completion.
+   *
    * This object can be cached and reused within the same [[GraphStageLogic]].
    */
   final protected def createAsyncCallback[T](handler: Procedure[T]): AsyncCallback[T] =
     getAsyncCallback(handler.apply)
 
+  private val concurrentAsyncCallback = ConcurrentHashMap.newKeySet[ConcurrentAsyncCallback[_]]().asScala
```

Comment:
Plural. I feel very uneasy about the wrapper, as `JSetWrapper` extends `SetLike` which does operations that are potentially not thread safe, I can't see that we are using such, but can we just use the Java API instead just to be sure?

Reply:
Better naming of this. What is it? `asyncCallbacksInProgess`? `pendingAsyncCallbacks`?

Reply:
these are `asyncCallbacksInProgess` as we are calling onStart and onStop to keep state up to date